composition-js/src/merging/merge.ts

```diff
@@ -3606,4 +3689,344 @@ class Merger {
+                  + `and/or @policy auth requirements to access the transitive interface "${parentType}" data from ${authRequirements.directive} selection set.`;
+              throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+            }
+            // re-check the implementation auth condition
+            this.verifyAuthOnFieldSelection(implType, selection, authRequirements);
+          }
+        }
+
+        if (parentType instanceof ObjectType) {
+          this.verifyAuthOnFieldSelection(parentType, selection, authRequirements);
+        }
+
+        if (selection.selectionSet) {
+          this.verifyAuthRequirementsOnSelectionSet(authRequirements, selection.selectionSet);
+        }
+      } else {
+        const condition = selection.element.typeCondition;
+        if (condition) {
+          const requirementsOnCondition = this.authRequirementsOnElement(condition);
+          if (!authRequirements.satisfies(requirementsOnCondition)) {
+            const msg = `Field "${authRequirements.fieldCoordinate}" does not specify necessary @authenticated, @requiresScopes `
+                + `and/or @policy auth requirements to access the transitive data in inline fragment type condition "${condition}" from ${authRequirements.directive} selection set.`;
+            throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+          }
+        }
+        this.verifyAuthRequirementsOnSelectionSet(authRequirements, selection.selectionSet);
+      }
+    }
+  }
+
+  private verifyAuthOnFieldSelection(parentType: ObjectType, selection: FieldSelection, authRequirements: AuthRequirements) {
+    const field = parentType.field(selection.element.name);
+    assert(field, 'field exists in the schema');
+    const fieldAuthReqs = this.authRequirementsOnElement(field);
+    const returnType = baseType(field.type!);
+    const fieldReturnAuthReqs = this.authRequirementsOnElement(returnType);
+
+    if (!authRequirements.satisfies(fieldAuthReqs) || !authRequirements.satisfies(fieldReturnAuthReqs)) {
+      const msg = `Field "${authRequirements.fieldCoordinate}" does not specify necessary @authenticated, @requiresScopes `
+          + `and/or @policy auth requirements to access the transitive field "${field.coordinate}" data from ${authRequirements.directive} selection set.`;
+      throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+    }
+  }
+}
+
+class AuthRequirements {
+  fieldCoordinate: string;
+  directive: string;
+  type?: AuthRequirementsOnElement;
+  field?: AuthRequirementsOnElement;
+
+  constructor(coordinate: string, directive: string) {
+    this.fieldCoordinate = coordinate;
+    this.directive = directive;
+  }
+
+  satisfies(authOnElement: AuthRequirementsOnElement | undefined): boolean {
+    if (authOnElement) {
+      const typeSatisfiesRequirements = this.type && this.type.satisfies(authOnElement);
+      const fieldSatisfiesRequirements = this.field && this.field.satisfies(authOnElement);
```

Comment:
You want a DNF conjunction of `type`'s and `field`'s auth requirements.

E.g. if `type` has `scopes` of `[["A"]]` and `field` has `scopes` of `[["B"]]`, then it's fine for a field in `@requires` to have `scopes` `[["A", "B"]]`.

Reply:
fixed in `isSubset` check